PolyEngine/Core/Src/BTreeMap.hpp

```diff
@@ -0,0 +1,532 @@
+		using KVERef   = typename BTree::KVERef;
+		using NodeRef  = typename BTree::NodeRef;
+		class Entry;
+		class Iter;
+		class Keys;
+		class Values;
+	public:
+		static constexpr size_t B = Bfactor;
+
+	public:
+		BTreeMap() : root{new LeafNode(), 0}, len(0) {} //todo(vuko): should we delay the allocation until the first insertion?
+		BTreeMap(BTreeMap&& other) : root(other.root), len(other.len) { ::new(&other) BTreeMap(); } //note(vuko): due to the previous issue this allocates!
+		//todo(vuko): copy constructor? tbh I'd rather avoid implementing it
+		~BTreeMap() { this->Clear(); delete this->root.node; };
+
+	public:
+		BTreeMap& operator=(BTreeMap&& other) {
+			this->~BTreeMap();
+			this->root = other.root;
+			this->len = other.len;
+			::new(&other) BTreeMap();
+			return *this;
+		}
+
+	public:
+		Optional<V> Insert(const K&  key, const V&  value) { return this->insert(          key ,           value ); }
+		Optional<V> Insert(const K&  key,       V&& value) { return this->insert(          key , std::move(value)); }
+		Optional<V> Insert(      K&& key, const V&  value) { return this->insert(std::move(key),           value ); }
+		Optional<V> Insert(      K&& key,       V&& value) { return this->insert(std::move(key), std::move(value)); }
+
+		Optional<V> Remove(const K&  key) { return this->remove(key); }
+		Optional<V> Remove(      K&& key) { return this->remove(std::move(key)); }
+
+		Optional<V&> Get(const K&  key) { return this->get(key); }
+		Optional<V&> Get(      K&  key) { return this->get(std::move(key)); }
+
+		V& operator[](const K& key) { return this->Get(key).value(); }
+		size_t GetSize() { return this->len; };
+		bool IsEmpty()   { return this->GetSize() == 0; };
+
+		void Clear() {
+			KVERef current = BTree::first_leaf_edge(this->root.as_node_ref());
+			while (this->len) {
+				this->len -= 1;
+				if (current.idx < current.node_ref.node->len) {
+					current = KVERef{current.node_ref, current.idx + 1};
+					continue;
+				} else {
+					const NodeRef to_delete = current.node_ref;
+					current = current.node_ref.ascend().parent;
+					delete to_delete.node;
+				}
+
+				for (;;) {
+					if (current.idx < current.node_ref.node->len) {
+						current = BTree::first_leaf_edge(KVERef{current.node_ref, current.idx + 1}.descend());
+						break; //continue outer loop
+					} else {
+						const NodeRef to_delete = current.node_ref;
+						current = current.node_ref.ascend().parent;
+						delete to_delete.node->as_branch();
+					}
+				}
+			}
+
+			LeafNode* new_root_to_be = current.node_ref.node;
+
+			auto ascension = current.node_ref.ascend();
+			while (ascension.succeeded) {
+				KVERef node = ascension.parent;
+				ascension = node.node_ref.ascend();
+				delete node.node_ref.node->as_branch();
+			}
+
+			//reinitialize the root node
+			new_root_to_be->~LeafNode();
+			::new(new_root_to_be) LeafNode();
+
+			this->root = {new_root_to_be, 0};
+		}
+
+	public:
+		Entry Entry(const K& key) {
```

Comment:
As we speaked earlier. It would be much better to replace Entry with proper iterators and find method.